### ISSUE TITLE
Unclear specification, use of DAZ + FTZ, reimplementation.

### ISSUE THREAD

**Issue by l1mey112** [tevador/RandomX#305]
Hello all. I am working on an implementation of RandomX in JavaScript that will do light verification at high performance, no dataset yet. For a small background, most of the C++ code is being reimplemented in freestanding C, with a WASM JIT for both virtual machines, glued together with WASM and JS. All of the requirements of RandomX that are not "available" in JS can be circumvented, with minor performance hits.

I am asking for clarification on the specification.

> All floating point operations are rounded according to the current value of the `fprc` register (see Table 4.3.1). Due to restrictions on the values of the floating point registers, no operation results in `NaN` or a denormal number.

The range of values presented to the register F additive instructions can actually result in denormals (subtractions, etc), however this never actually comes into play because FTZ (flush to zero) and DAZ (denormals are zero) flags are set in the FPU. I can see this behavior in the x86 build of RandomX, I've looked in the other architectures but couldn't find anything conclusive.

The use of FTZ and DAZ isn't documented in the specification. Would an implementation need to set these? I am building a test suite for my FP arithmetic library using real values consumed by RandomX, and it seems that FTZ + DAZ has an effect.

---

I would also like to get in contact with the developers of RandomX through the proper channels to ask questions and ensure correctness with the original library as I release mine. It would be beneficial to have multiple implementations of this library. 

https://github.com/l1mey112/randomx.js

**Comment by SChernykh** [comment]
Group F registers don't get smaller than 10^4 by the absolute value: https://github.com/tevador/RandomX/blob/master/doc/design.md#251-floating-point-register-groups

If you have a counter-example where denormals actually happen, please post it here.

**Comment by l1mey112** [comment]
You are right. I used incorrect flags for the MXCSR, using correct flags below ended up with the correct result regardless.

```c++
constexpr uint32_t rx_mxcsr_default = 0x1f80;
//constexpr uint32_t rx_mxcsr_default = 0x9FC0; //Flush to zero, denormals are zero, default rounding mode, all exceptions disabled
```

As for the generation of subnormals I did test about 2 million data points, all normal values. For the floating point result of exactly zero, it came up in tests that the only way a zero can be generated is through the identity `x - x = 0`, that's where they all came from, never a `x - xclose = 0.subn`.

![FP heap test](https://github.com/user-attachments/assets/a0f3a6f7-aa3b-47a9-97e2-be78f818c141)

All zero values generated are perfectly rounded, never ending up in a subnormal value. Taking advantage of this in FP emulation did not break any of the 100 thousand sample points from RandomX (test above). This issue can be closed.

---

I do want to ask how I could get into proper contact with Tevador or any of there other RandomX developers through proper channels. Is there an IRC server they are present on?
